mobilecoind/src/db_crypto.rs

```diff
@@ -0,0 +1,184 @@
+    /// backend or without).
+    fn requires_password(&self) -> bool {
+        false
+    }
+
+    /// Encrypt data.
+    fn encrypt(&self, data: &[u8]) -> Result<Vec<u8>, DbCryptoError> {
+        Ok(data.to_vec())
+    }
+
+    /// Decrypt data.
+    fn decrypt(&self, data: &[u8]) -> Result<Vec<u8>, DbCryptoError> {
+        Ok(data.to_vec())
+    }
+}
+
+/// Aes crypto provider
+#[derive(Clone)]
+pub struct AesDbCryptoProvider {
+    encryption_key: Arc<Mutex<Option<Vec<u8>>>>,
+}
+
+impl Default for AesDbCryptoProvider {
+    fn default() -> Self {
+        Self {
+            encryption_key: Arc::new(Mutex::new(None)),
+        }
+    }
+}
+
+impl DbCryptoProvider for AesDbCryptoProvider {
+    /// Sets the password to use.
+    fn set_password(&self, password: &[u8]) -> Result<(), DbCryptoError> {
+        let mut encryption_key = self.encryption_key.lock().expect("muted poisoned");
+        if encryption_key.is_some() {
+            Err(DbCryptoError::EncryptionKeyAlreadySet)
+        } else if password.len() != 32 {
+            Err(DbCryptoError::EncryptionInvalidKeyLength)
+        } else {
+            *encryption_key = Some(password.to_vec());
+            Ok(())
+        }
+    }
+
+    /// Clear the password.
+    fn clear_password(&self) -> Result<(), DbCryptoError> {
+        let mut encryption_key = self.encryption_key.lock().expect("muted poisoned");
+        *encryption_key = None;
+        Ok(())
+    }
+
+    /// Whether a password is needed or not (allows us to tell if we're running with an encryption
+    /// backend or without).
+    fn requires_password(&self) -> bool {
+        true
+    }
+
+    /// Encrypt data.
+    fn encrypt(&self, plaintext_bytes: &[u8]) -> Result<Vec<u8>, DbCryptoError> {
+        let (key, nonce) = self.expand_password()?;
+
+        let key_arr = GenericArray::from_slice(&key[..]);
+        let nonce_arr = GenericArray::from_slice(&nonce[..]);
+
+        // Get cipher from hash bytes
+        let cipher = Aes256Gcm::new(key_arr);
+
+        Ok(cipher.encrypt(nonce_arr, &plaintext_bytes[..])?)
+    }
+
+    /// Decrypt data.
+    fn decrypt(&self, ciphertext: &[u8]) -> Result<Vec<u8>, DbCryptoError> {
+        let (key, nonce) = self.expand_password()?;
+        let key_arr = GenericArray::from_slice(&key[..]);
+        let nonce_arr = GenericArray::from_slice(&nonce[..]);
+
+        let cipher = Aes256Gcm::new(key_arr);
+        Ok(cipher.decrypt(nonce_arr, ciphertext)?)
+    }
+}
+
+impl AesDbCryptoProvider {
+    /// Get the encryption key if available, or an error if one is needed and hasn't been provided
+    /// yet.
+    fn get_encryption_key(&self) -> Result<Vec<u8>, DbCryptoError> {
+        let encryption_key = self.encryption_key.lock().expect("muted poisoned");
+        encryption_key
+            .clone()
+            .ok_or(DbCryptoError::MissingEncryptionKey)
+    }
+
+    /// TODO
+    fn expand_password(&self) -> Result<([u8; 32], [u8; 12]), DbCryptoError> {
```

Comment:
it might be a little cleaner if you return GenericArray types here, but it doesn't matter much